src/main/java/com/adq/jenkins/xmljobtodsl/dsl/strategies/custom/DSLCleanupPatternStrategy.java

```diff
@@ -0,0 +1,93 @@
+package com.adq.jenkins.xmljobtodsl.dsl.strategies.custom;
+
+import com.adq.jenkins.xmljobtodsl.dsl.strategies.DSLObjectStrategy;
+import com.adq.jenkins.xmljobtodsl.parsers.PropertyDescriptor;
+import java.util.ArrayList;
+import java.util.List;
+
+/*
+Strategy for preBuildCleanup attribute, to correctly render includePattern and excludePattern, which is in the XML as:
+<hudson.plugins.ws__cleanup.PreBuildCleanup plugin="ws-cleanup@0.43">
+            <patterns>
+                <hudson.plugins.ws__cleanup.Pattern>
+                    <pattern>results/*</pattern>
+                    <type>INCLUDE</type>
+                </hudson.plugins.ws__cleanup.Pattern>
+            </patterns>
+            <deleteDirs>false</deleteDirs>
+            <cleanupParameter>test</cleanupParameter>
+        </hudson.plugins.ws__cleanup.PreBuildCleanup>
+
+ The value within type determines the method name: either "includePattern" or "excludePattern"
+ "pattern" is then inserted as the value for that method
+  Returns the dsl as:
+   preBuildCleanup {
+			includePattern("test")
+			deleteDirectories(false)
+			cleanupParameter("test")
+		}
+ */
+
+public class DSLCleanupPatternStrategy extends DSLObjectStrategy {
+
+    public DSLCleanupPatternStrategy(int tabs, PropertyDescriptor propertyDescriptor, String name) {
+        this(tabs, propertyDescriptor, name, true);
+    }
+
+    public DSLCleanupPatternStrategy(int tabs, PropertyDescriptor propertyDescriptor, String name, boolean shouldInitChildren) {
+        super(tabs, propertyDescriptor, name, shouldInitChildren);
+
+        List<PropertyDescriptor> children = propertyDescriptor.getProperties();
+        List<PropertyDescriptor> leftoverProps = new ArrayList<>();
+        List<PropertyDescriptor> patternProps = new ArrayList<>();
+
+        // separating the properties nested under patterns to handle later
+        for (PropertyDescriptor child : children) {
+            if (!child.getName().equals("patterns")) {
+                leftoverProps.add(child);
+            }
+            if (child.getName().equals("patterns")) {
+                patternProps.add(child);
+            }
+        }
+
+        // extracting type and pattern property descriptors
+        PropertyDescriptor typeChild = null;
+        PropertyDescriptor patternChild = null;
+        for (PropertyDescriptor prop : patternProps) {
+            if (prop.getName().equals("patterns")) {
+                for (PropertyDescriptor innerProp : prop.getProperties()) {
+                    for (PropertyDescriptor nestedProp : innerProp.getProperties()) {
+                        if (nestedProp.getName().equals("pattern")) {
+                            patternChild = nestedProp;
+                        }
+                        if (nestedProp.getName().equals("type")) {
+                            typeChild = nestedProp;
+                        }
+                    }
+                }
+            }
```

Comment:
Brutal! But for now let's roll with it

Reply:
Yes, because those are so nested and I started at the parent, this is ugly! But I figured get it working and then iterate on that.

Reply:
The only other thing I could think of would be recursion to find the correct nodes ie https://github.com/tinyspeck/xml-job-to-job-dsl-plugin/blob/master/src/main/java/com/adq/jenkins/xmljobtodsl/dsl/strategies/DSLMethodStrategy.java#L96

except instead of returning ALL children just return children named pattern and type